In [1]:
# read a parquet file and print the schema
import pandas as pd

filename = 'data/test/yellow_tripdata_2025-01.parquet'
data = pd.read_parquet(filename)
print(data.dtypes)
print(data.head())

VendorID                          int32
tpep_pickup_datetime     datetime64[us]
tpep_dropoff_datetime    datetime64[us]
passenger_count                 float64
trip_distance                   float64
RatecodeID                      float64
store_and_fwd_flag               object
PULocationID                      int32
DOLocationID                      int32
payment_type                      int64
fare_amount                     float64
extra                           float64
mta_tax                         float64
tip_amount                      float64
tolls_amount                    float64
improvement_surcharge           float64
total_amount                    float64
congestion_surcharge            float64
Airport_fee                     float64
dtype: object
   VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0         1  2025-01-01 00:18:38   2025-01-01 00:26:59              1.0   
1         1  2025-01-01 00:32:40   2025-01-01 00:35:13              1.0   
2

In [5]:
import glob
import os

parquet_folder = 'data/train'
all_files = glob.glob(os.path.join(parquet_folder, 'yellow_tripdata_2024-*.parquet'))
df_list = [pd.read_parquet(file) for file in all_files]
df = pd.concat(df_list, ignore_index=True)

In [6]:
df.describe()

VendorID        tpep_pickup_datetime       tpep_dropoff_datetime  \
count  4.116972e+07                    41169720                    41169720   
mean   1.764232e+00  2024-07-06 10:01:25.051413  2024-07-06 10:18:53.125909   
min    1.000000e+00         2002-12-31 16:46:07         2002-12-31 17:24:07   
25%    2.000000e+00  2024-04-06 20:07:27.500000         2024-04-06 20:23:52   
50%    2.000000e+00         2024-07-03 23:35:15         2024-07-03 23:50:35   
75%    2.000000e+00         2024-10-08 17:33:35  2024-10-08 17:53:27.250000   
max    7.000000e+00         2026-06-26 23:53:12         2026-06-27 20:59:10   
std    4.258568e-01                         NaN                         NaN   

       passenger_count  trip_distance    RatecodeID  PULocationID  \
count     3.707849e+07   4.116972e+07  3.707849e+07  4.116972e+07   
mean      1.333931e+00   4.976101e+00  2.322150e+00  1.642428e+02   
min       0.000000e+00   0.000000e+00  1.000000e+00  1.000000e+00   
25%       1.000000e+00   1.010000e+00  1.000000e+00  1.320000e+02   
50%       1.000000e+00   1.760000e+00  1.000000e+00  1.610000e+02   
75%       1.000000e+00   3.360000e+00  1.000000e+00  2.330000e+02   
max       9.000000e+00   3.986086e+05  9.900000e+01  2.650000e+02   
std       8.158242e-01   4.192305e+02  1.092805e+01  6.434069e+01   

       DOLocationID  payment_type   fare_amount         extra       mta_tax  \
count  4.116972e+07  4.116972e+07  4.116972e+07  4.116972e+07  4.116972e+07   
mean   1.634475e+02  1.107259e+00  1.926851e+01  1.385953e+00  4.797774e-01   
min    1.000000e+00  0.000000e+00 -2.261200e+03 -9.250000e+00 -5.000000e-01   
25%    1.130000e+02  1.000000e+00  9.300000e+00  0.000000e+00  5.000000e-01   
50%    1.620000e+02  1.000000e+00  1.350000e+01  1.000000e+00  5.000000e-01   
75%    2.340000e+02  1.000000e+00  2.260000e+01  2.500000e+00  5.000000e-01   
max    2.650000e+02  5.000000e+00  3.355444e+05  6.599000e+01  4.130000e+01   
std    6.960009e+01  6.515108e-01  7.671984e+01  1.815878e+00  1.301830e-01   

         tip_amount  tolls_amount  improvement_surcharge  total_amount  \
count  4.116972e+07  4.116972e+07           4.116972e+07  4.116972e+07   
mean   3.307884e+00  5.615266e-01           9.629934e-01  2.783281e+01   
min   -3.000000e+02 -1.406300e+02          -1.000000e+00 -2.265450e+03   
25%    0.000000e+00  0.000000e+00           1.000000e+00  1.575000e+01   
50%    2.600000e+00  0.000000e+00           1.000000e+00  2.100000e+01   
75%    4.250000e+00  0.000000e+00           1.000000e+00  3.060000e+01   
max    9.999900e+02  1.702880e+03           2.000000e+00  3.355509e+05   
std    4.090523e+00  2.240545e+00           2.550554e-01  7.805359e+01   

       congestion_surcharge   Airport_fee  
count          3.707849e+07  3.707849e+07  
mean           2.232144e+00  1.470060e-01  
min           -2.500000e+00 -1.750000e+00  
25%            2.500000e+00  0.000000e+00  
50%            2.500000e+00  0.000000e+00  
75%            2.500000e+00  0.000000e+00  
max            2.520000e+00  1.750000e+00  
std            8.746528e-01  5.020407e-01

In [7]:
df.isnull().sum()

VendorID                       0
tpep_pickup_datetime           0
tpep_dropoff_datetime          0
passenger_count          4091232
trip_distance                  0
RatecodeID               4091232
store_and_fwd_flag       4091232
PULocationID                   0
DOLocationID                   0
payment_type                   0
fare_amount                    0
extra                          0
mta_tax                        0
tip_amount                     0
tolls_amount                   0
improvement_surcharge          0
total_amount                   0
congestion_surcharge     4091232
Airport_fee              4091232
dtype: int64

In [8]:
parquet_folder = 'data/train'
all_files = glob.glob(os.path.join(parquet_folder, '*.parquet'))
df_list = [pd.read_parquet(file) for file in all_files]
df2 = pd.concat(df_list, ignore_index=True)

In [9]:
df2.isnull().sum()

VendorID                        0
lpep_pickup_datetime     41169720
lpep_dropoff_datetime    41169720
store_and_fwd_flag        4115560
RatecodeID                4115560
PULocationID                    0
DOLocationID                    0
passenger_count           4115560
trip_distance                   0
fare_amount                     0
extra                           0
mta_tax                         0
tip_amount                      0
tolls_amount                    0
ehail_fee                41829938
improvement_surcharge           0
total_amount                    0
payment_type                24328
trip_type                41194130
congestion_surcharge      4115560
tpep_pickup_datetime       660218
tpep_dropoff_datetime      660218
Airport_fee               4751450
dtype: int64

In [3]:
filename = 'data/test/green_tripdata_2025-01.parquet'
data = pd.read_parquet(filename)
data.describe()

VendorID        lpep_pickup_datetime       lpep_dropoff_datetime  \
count  48326.000000                       48326                       48326   
mean       1.870215  2025-01-17 04:19:01.965753  2025-01-17 04:38:59.277448   
min        1.000000         2024-12-25 23:13:15         2024-12-25 23:13:17   
25%        2.000000  2025-01-09 14:00:31.500000  2025-01-09 14:26:33.500000   
50%        2.000000  2025-01-16 19:46:55.500000  2025-01-16 20:01:19.500000   
75%        2.000000  2025-01-24 17:47:35.500000         2025-01-24 18:03:52   
max        2.000000         2025-02-05 18:46:24         2025-02-05 19:11:47   
std        0.336070                         NaN                         NaN   

         RatecodeID  PULocationID  DOLocationID  passenger_count  \
count  46490.000000  48326.000000  48326.000000     46490.000000   
mean       1.184771     94.094545    142.409014         1.278103   
min        1.000000      3.000000      1.000000         0.000000   
25%        1.000000     74.000000     74.000000         1.000000   
50%        1.000000     75.000000    140.000000         1.000000   
75%        1.000000     97.000000    230.000000         1.000000   
max       99.000000    265.000000    265.000000         9.000000   
std        1.442502     54.968061     77.251550         0.937178   

       trip_distance   fare_amount         extra       mta_tax    tip_amount  \
count   48326.000000  48326.000000  48326.000000  48326.000000  48326.000000   
mean       21.532410     16.762466      0.932324      0.602776      2.481859   
min         0.000000   -113.000000     -5.000000     -0.500000     -0.900000   
25%         1.100000      9.300000      0.000000      0.500000      0.000000   
50%         1.740000     13.500000      0.000000      0.500000      2.070000   
75%         2.940000     19.100000      2.500000      0.500000      3.690000   
max     84731.570000    336.200000      7.500000      4.250000    252.050000   
std       990.646907     13.308342      1.348587      0.357366      3.213612   

       tolls_amount  ehail_fee  improvement_surcharge  total_amount  \
count  48326.000000        0.0           48326.000000  48326.000000   
mean       0.177461        NaN               0.988346     22.634242   
min        0.000000        NaN              -1.000000   -114.000000   
25%        0.000000        NaN               1.000000     13.705000   
50%        0.000000        NaN               1.000000     18.750000   
75%        0.000000        NaN               1.000000     26.465000   
max       48.760000        NaN               1.000000    371.400000   
std        1.192984        NaN               0.130402     15.435061   

       payment_type     trip_type  congestion_surcharge  
count  46490.000000  46483.000000           46490.00000  
mean       1.266423      1.038788               0.85306  
min        1.000000      1.000000              -2.75000  
25%        1.000000      1.000000               0.00000  
50%        1.000000      1.000000               0.00000  
75%        2.000000      1.000000               2.75000  
max        5.000000      2.000000               2.75000  
std        0.471842      0.193092               1.27190

In [4]:
data.isnull().sum()

VendorID                     0
lpep_pickup_datetime         0
lpep_dropoff_datetime        0
store_and_fwd_flag        1836
RatecodeID                1836
PULocationID                 0
DOLocationID                 0
passenger_count           1836
trip_distance                0
fare_amount                  0
extra                        0
mta_tax                      0
tip_amount                   0
tolls_amount                 0
ehail_fee                48326
improvement_surcharge        0
total_amount                 0
payment_type              1836
trip_type                 1843
congestion_surcharge      1836
dtype: int64

In [21]:
data['Airport_fee'].value_counts()

Airport_fee
 0.00    2706446
 1.75     218203
-1.75      10411
 1.25          8
 5.00          7
 0.75          1
 6.75          1
Name: count, dtype: int64

In [19]:
# summarize extra column
print(data['extra'].describe())

# get p90 of extra column
print(data['extra'].quantile(0.9))

count    3.475226e+06
mean     1.317737e+00
std      1.861509e+00
min     -7.500000e+00
25%      0.000000e+00
50%      0.000000e+00
75%      2.500000e+00
max      1.500000e+01
Name: extra, dtype: float64
4.25


In [20]:
# summaruze mta tax
print(data['mta_tax'].describe())

count    3.475226e+06
mean     4.780991e-01
std      1.374623e-01
min     -5.000000e-01
25%      5.000000e-01
50%      5.000000e-01
75%      5.000000e-01
max      1.050000e+01
Name: mta_tax, dtype: float64


In [2]:
# check if there are any missing values in each column
print(data.isnull().sum())

VendorID                      0
tpep_pickup_datetime          0
tpep_dropoff_datetime         0
passenger_count          540149
trip_distance                 0
RatecodeID               540149
store_and_fwd_flag       540149
PULocationID                  0
DOLocationID                  0
payment_type                  0
fare_amount                   0
extra                         0
mta_tax                       0
tip_amount                    0
tolls_amount                  0
improvement_surcharge         0
total_amount                  0
congestion_surcharge     540149
Airport_fee              540149
dtype: int64


In [5]:
# number of rows and columns
print(data.shape)

(3475226, 19)


In [6]:
# unique pickup and dropoff locations
print(data['PULocationID'].nunique())
print(data['DOLocationID'].nunique())

261
260


In [8]:
# unique pickup and dropoff locations combined
locations = pd.concat([data['PULocationID'], data['DOLocationID']])
print(locations.nunique())

262


In [12]:
dec_data = pd.read_parquet('yellow_tripdata_2024-12.parquet')
print(dec_data.shape)
locations = pd.concat([dec_data['PULocationID'], dec_data['DOLocationID']])
print(locations.nunique())

(3668371, 19)
261


In [13]:
apr_data = pd.read_parquet('yellow_tripdata_2024-04.parquet')
print(apr_data.shape)
locations = pd.concat([apr_data['PULocationID'], apr_data['DOLocationID']])
print(locations.nunique())

(3514289, 19)
262


A, B, C, D

PU, DO
A, B
B, A
..
C, D
D, C

# one-hot encoding for PU and DO separately
PU_A, PU_B, PU_C, PU_D, DO_A, DO_B, DO_C, DO_D - n * 2 columns: 260 * 2: less  number of columns to deal with
1,      0,    0,    0,    0,    0, 1, 0, 0


Routes - 16 routes - n * n
A->B
B->A
...
C->D
D->C

Route_AB, Route_BA, ... Route_CD, Route_DC - n * n columns: 260 * 260: too many columns but relationship is maintained
1, 0,...0
0, 1, ...0
....
0, 0, ...., 1, 0



In [22]:
!pip install geopandas

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 15.1/15.1 MB 18.3 MB/s eta 0:00:0000:0100:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.7/4.7 MB 17.3 MB/s eta 0:00:00a 0:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 17.3 MB/s eta 0:00:00


In [5]:
import geopandas as gpd
 
# Load the shapefile
zones = gpd.read_file("taxi_zones/taxi_zones.shp")
print(zones.head())
 
# Compute centroids
zones['longitude'] = zones['geometry'].centroid.x
zones['latitude'] = zones['geometry'].centroid.y

   OBJECTID  Shape_Leng  Shape_Area                     zone  LocationID  \
0         1    0.116357    0.000782           Newark Airport           1   
1         2    0.433470    0.004866              Jamaica Bay           2   
2         3    0.084341    0.000314  Allerton/Pelham Gardens           3   
3         4    0.043567    0.000112            Alphabet City           4   
4         5    0.092146    0.000498            Arden Heights           5   

         borough                                           geometry  
0            EWR  POLYGON ((933100.918 192536.086, 933091.011 19...  
1         Queens  MULTIPOLYGON (((1033269.244 172126.008, 103343...  
2          Bronx  POLYGON ((1026308.77 256767.698, 1026495.593 2...  
3      Manhattan  POLYGON ((992073.467 203714.076, 992068.667 20...  
4  Staten Island  POLYGON ((935843.31 144283.336, 936046.565 144...  


In [7]:
zones.shape

(263, 9)